lib/install/opsuser/opsuser.go

```diff
@@ -174,24 +198,39 @@ func (mgr *RBACManager) setupPermissions(ctx context.Context) ([]rbac.ResourcePe
 		return nil, errors.Errorf("Ops-User: RBACManager, Unable to find Datastores: %s", err.Error())
 	}
 
-	// Loop over Datastores
+	// Add datastores
 	for _, ref := range dsNameToRef {
-		resourceDescs = append(resourceDescs, ResourceDesc{rbac.Datastore, ref})
+		resourceDescs = append(resourceDescs, resourceDesc{rbac.Datastore, ref})
 	}
 
-	// Loop over Networks
+	// Add networks
 	for _, network := range mgr.configSpec.Network.ContainerNetworks {
 		netRef := &types.ManagedObjectReference{}
 		netRef.FromString(network.ID)
 		if netRef.Type == "" || netRef.Value == "" {
 			return nil, errors.Errorf("Ops-User: RBACManager, Unable to build Bridged Network MoRef: %s", network.ID)
 		}
-		resourceDescs = append(resourceDescs, ResourceDesc{rbac.Network, *netRef})
+		resourceDescs = append(resourceDescs, resourceDesc{rbac.Network, *netRef})
+	}
+
+	// In a DRS environment, add the endpoint role to the resource pool(s). In a non-DRS
+	// environment, add the endpoint role to the cluster.
+	if mgr.session.DRSEnabled != nil && *mgr.session.DRSEnabled {
+		for _, rPoolRef := range mgr.configSpec.ComputeResources {
+			resourceDescs = append(resourceDescs, resourceDesc{rbac.Endpoint, rPoolRef})
+		}
+	} else {
+		resourceDescs = append(resourceDescs, resourceDesc{rbac.Endpoint, cluster.Reference()})
```

Comment:
I'm not sure how this is intended to work. As I understand it, vSphere only allows a user to have a single role on a given entity. We've already assigned a role to the cluster on line 192.

Reply:
Looking into this.

Reply:
Good question - so on line 192, we assign the `vic-vch-datastore` role (https://github.com/vmware/vic/blob/master/lib/install/opsuser/opsuser_conf.go#L62-L72) where `Propagate` is `true` - which would apply those permissions to the cluster's sub-objects. Then, on line 223, we apply the `vic-vch-endpoint` role to the cluster.

I added debug logs to retrieve the cluster's permissions after we've added both roles to it - and it looks like only the role that was added last (`vic-vch-endpoint` in our case) is kept. Then, I removed the logic on line 192 where we add `vic-vch-datastore` to the cluster (and its sub-entities) but then `vic-machine create --container-network=...` would fail (time out) at the `vic/lib/install/management.(*Dispatcher).ensureApplianceInitializes` function with:

```
ERRO[0221] vic/lib/install/management.(*Dispatcher).ensureApplianceInitializes: Create error: context deadline exceeded
ERRO[0221] Failed to obtain IP address for client interface
...
INFO[0221] 		Run vic-machine-linux inspect to find API connection command and run the command if ip address is ready 
INFO[0221] 		If command succeeds, VCH is started. If command fails, VCH failed to install - see documentation for troubleshooting. 
ERRO[0221] vic/cmd/vic-machine/create.(*Create).Run.func3: Create error: context deadline exceeded
...
ERRO[0221] vic-machine-linux create failed: Creating VCH exceeded time limit of 3m0s. Please increase the timeout using --timeout to accommodate for a busy vSphere target
```

A subsequent `vic-machine inspect` fails with `No client IP address assigned`. It's worth noting that a `vic-machine create` without any container networks succeeds here.

Per my understanding, the logic on line 192 is necessary to assign those permissions to the cluster's sub-objects (most likely the network devices since specifying container networks in the `vic-machine create` command fails if these permissions aren't present).

Also, while working on https://github.com/vmware/vic/issues/7053, I opened https://github.com/vmware/vic/issues/7606 as a follow-on issue for refining our permissions and roles logic.

Reply:
When the `vic-vch-datastore` role is overridden with the `vic-vch-endpoint` role all privileges will be lost, including propagated privileges (as propagation is handled by vSphere, not us).

In other words, I'd expect the failure you see when removing line 192 to eventually occur even if the line is not removed (if not during the initial create operation, then when performing container operations or during a reconfigure).

I expect that you'll need to define a new role, such as `vic-vch-endpoint-datastore`, that is the union of the `vic-vch-datastore` and `vic-vch-endpoint` roles and use that here. Essentially, this is our workaround for vSphere's single-role limitation: create an additional single role representing the two we want to apply.

To avoid a maintenance nightmare around keeping these in sync, I'd suggest explicitly using a union operation as a part of the role definition.

I.e., something like:
```go
var RoleEndpointDatastore = types.AuthorizationRole{
	Name:      "endpoint-datastore",
	Privilege: append(RoleEndpoint.Privilege, RoleDataStore.Privilege...),
}
```

(Eventually, we might refactor this code to create these "union roles" dynamically, but I think that's a larger change and not something we should block ROBO on.)

Reply:
When the `vic-vch-datastore` role is overridden with the `vic-vch-endpoint` role all privileges will be lost, including propagated privileges (as propagation is handled by vSphere, not us).

In other words, I'd expect the failure you see when removing line 192 to eventually occur even if the line is not removed (if not during the initial create operation, then when performing container operations or during a reconfigure).

I expect that you'll need to define a new role, such as `vic-vch-endpoint-datastore`, that is the union of the `vic-vch-datastore` and `vic-vch-endpoint` roles and use that here. Essentially, this is our workaround for vSphere's single-role limitation: create an additional single role representing the two we want to apply.

To avoid a maintenance nightmare around keeping these in sync, I'd suggest explicitly using a union operation as a part of the role definition.

I.e., something like:
```go
var RoleEndpointDatastore = types.AuthorizationRole{
	Name:      "endpoint-datastore",
	Privilege: append(RoleEndpoint.Privilege, RoleDataStore.Privilege...),
}
```

(Eventually, we might refactor this code to create these "union roles" dynamically, but I think that's a larger change and not something we should block ROBO on.)

Reply:
As a concrete example of something that breaks when the `vic-vch-endpoint` role is applied to the Cluster and the `vic-vch-datastore` is applied only to the datastores: the operations user will not have the `Host.Config.SystemManagement` privilege on the hosts within the cluster.

This can be seen by visiting the "Permissions" tab of the host via the Flex UI; it will show that the operations user inherits the `vic-vch-endpoint` role defined on the cluster and the `vic-vch-datacenter` role defined on the datacenter, but does not inherit the `vic-vch-datastore` role defined on the datastores and, therefore, does not inherit the `Host.Config.SystemManagement` privilege.

On my system, I can successfully deploy a VCH, start a container, and download a container log bundle from vicadmin when using an operations user that's misconfigured in this way. However, the container log files in that log bundle are all empty.

When this happens, and error like this is written to the vicadmin log:
> ```Apr 11 2018 22:37:05.518Z WARN  error reading pensive_brattain-1b9b253aab8a/output.log({%!s(*object.Datastore=&{{/dc1/datastore/sharedVmfs-0 0xc42032c2c0 {Datastore datastore-49}} /dc1}) 1b9b253aab8ab46f5dbe917c473e80040821a6be1634934785d033843499a121/output.log %!s(*context.valueCtx=&{0xc420014430 {} 0xc42026a840})}): ServerFaultCode: Permission to perform this operation was denied.```

Similar errors are seen for other files, although I'm not sure if these are significant or related:
> ```Apr 11 2018 22:37:05.176Z WARN  error reading appliance/tether.debug({%!s(*object.Datastore=&{{/dc1/datastore/sharedVmfs-0 0xc42032c2c0 {Datastore datastore-49}} /dc1}) robo-test-3/tether.debug %!s(*context.valueCtx=&{0xc420014430 {} 0xc4206e5f80})}): ServerFaultCode: Permission to perform this operation was denied.```

> ```Apr 11 2018 22:37:05.579Z WARN  error reading hostd/HostSystem:host-55 @ /dc1/host/cluster1/10.162.9.135({%!s(*session.Session=&{0xc42054e320 0xc4203076b0 0xc4202b5640 0xc42026bb40 0xc4204d9a90 <nil> 0xc420279800 0xc42027f340 0xc42027f340 0xc4204d7140 0xc4204bb43f}) hostd %!s(*object.HostSystem=&{{/dc1/host/cluster1/10.162.9.135 0xc42032c2c0 {HostSystem host-55}}})}): ServerFaultCode: Permission to perform this operation was denied.```

When I perform the same set of operations and then manually adjust role assignment to ensure that the operations user has the `Host.Config.SystemManagement` privilege on all hosts in the cluster, I receive a fully-populated log bundle.

Reply:
Thanks! I've created https://github.com/vmware/vic/issues/7721 to track this.